### ISSUE TITLE
Confusing application naming and version numbering

### ISSUE THREAD

**Issue by jedfrechette** [pyblish/pyblish-win#10]
This is probably a more general issue with the way release naming and numbering is handled but it is most easily illustrated with the Windows installer so I'm reporting it here. Consider the following series of user interactions:
1. User downloads installer named: pyblish-1.2.4-x64-setup.exe
2. User installs application described as "Pyblish 1.2.4" within installer.
3. User checks Programs and Features from the Control Panel and confirms that "Pyblish" version 1.2.4 has been installed.
4. User goes to the Start Menu and launches "Pyblish".
5. User opens "Pyblish" from it's tray icon.
6. User goes to the Terminal tab and finds they have Pyblish 1.2.1.

I understand that this is because in steps 1-3 the user is actually interacting with pyblish-win, which is at version 1.2.4, while only in step 6 do they interact with the real pyblish that is still at version 1.2.1. Add in things like searching pypi and finding that the latest version of pyblish is 1.2.1 and this loose terminology seems like a recipe for confusion.

**Comment by mottosso** [comment]
It's quite general yes, but necessary to bring up and I think you're right.

I've had an idea in mind for the past few weeks now, which is to put Pyblish X more at the forefront of distribution and have it's version set the tone for what version Pyblish is currently at for the end-user.

At the moment, each distribution (only Windows so far) builds upon Pyblish X - whenever it updates, they update too. And at the moment, they each increment their own version. But an alternative might be to simply "forward" the version of `pyblish-x` and to remove the notion of a version from `pyblish-win`. The only real problem I can see with that is if `pyblish-win` needs to update individually, which it can, for example to fix a bug or add to its list of executables. We wouldn't want to have to update `pyblish-x` due to an update in one of it's "subclasses".

The core library could then be represented, in the Terminal for example, as `core`, or `pyblish-core`, alongside the other libraries such as `pyblish-rpc`. If at all?

I'm interested in hearing your thoughts about this, or perhaps if you have something else in mind?

**Comment by jedfrechette** [comment]
Looking at this from the outside I wonder what Pyblish refers to. Is Pyblish a small module that forms the base other modules are built on or is it a suite of tools? Looking at another example from the Python ecosystem, is it Traits [1] or is it the Enthought Tool Suite [2]?

If it's a small base module then I think the current project names make sense, although `pyblish-x` is perhaps not the most descriptive name. If it is a suite of tools then perhaps `pyblish` should be renamed to `pyblish-core` and `pyblish-x` renamed to `pyblish`. The way you've packaged `pyblish-win` and you're comments above makes me think Pyblish is a suite in which case I would choose the later. You could certainly keep the name Pyblish X but I think renaming `pyblish-x` to `pyblish` simplifies and reinforces the Pyblish branding. A user looking for Pyblish then installs Pyblish, whether via a standalone installer, a package management system, or manually and get's the full suite with all of the sub-components, which is probably what they want.

Instead of a Python package, `pyblish-win` looks like a distribution for a specific platform to me. Rather than `pyblish-linux`, I guess I would expect to see other distributions targeting specific platforms, e.g. `pyblish-deb`, `pyblish-rpm`, or `pyblish-conda`. For versioning these distributions you could do something similar to what those package management systems do where package version numbers are derived by combining the upstream version number and an optional packaging revision number. Following the Debian [3] or Redhat convention, instead of the current `pyblish-win` version being 1.2.4 it would be 1.2.1-2 if `pyblish` is considered to be the upstream version or 1.2.2-2 if `pyblish-x` is upstream.

It might also be worth considering a different naming convention for these distributions. In the `pyblish` wiki you're very specific about using the `pyblish-*` pattern for Extentions and Integrations. Looking at the list of Pyblish repos I might guess that, instead of a distribution targeting the Windows platform, `pyblish-win` is actually an Integration of Pyblish in to Windows Explorer.

Of course all of this needs to be balanced against backwards compatibility which you're certainly better positioned to assess than I am.

Finally, a couple other questions I've had while looking at all of this, which might make it more clear how these interactions should look: With the current setup would a user ever write `import pyblish_x` or `import pyblish_win`? How about `pip install pyblish-win`?

[1] https://github.com/enthought/traits
[2] https://github.com/enthought/ets
[3] https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-Version

**Comment by mottosso** [comment]
Great, great observations @jedfrechette. Thanks for taking the time to think about this.

I'll try not to defend the way things are currently, because it has if anything been "grown" as opposed to designed to look and function this way, and because I am currently looking to delegate more of the overall design to the community; which is a delicate balance between "anything goes" and "don't touch". But I will at least try and provide some motivation behind some of the decisions so you get a fuller picture.

Firstly, we need to separate between what Pyblish is to the **end-user**, and what it is to a **maintainer**.

### To the user

To the user, the main thing I've tried to keep in mind is that GitHub, and all of the Pyblish repositories stored here, is an implementation detail of Pyblish. From the user's perspective, I'd like there to be single installation method fit for his platform and a single API. E.g. for a Windows user, this would be the `setup.exe` file and for a Linux user, a `.deb` or `.rpm` package and so forth.

Qt has been a primary inspiration for this. It consists of many modules, some of which work in isolation whereas others do not. To the user, Qt is a single library, with a single (albeit large) API. This is how I see Pyblish, and the fact that there are many Git repositories is more of a maintenance technique meant for those developing Pyblish.

Perhaps this fact isn't prominent enough, and ideally I would like there to be a more prominent webpage to illustrate this, such that GitHub can take a backseat. The reason GitHub is so heavily used today is simply because it makes editing things (like the Wiki) easier than editing a bespoke webpage. The forum has since taken a lot of that work off it's hands, which I think has moved things in the right direction.

> Is Pyblish a small module that forms the base other modules are built on or is it a suite of tools?

Technically, the [pyblish](https://github.com/pyblish/pyblish) git repository is a small module upon which all other modules are built. There is only one tool, which is Pyblish. At least this is the impression I would like to convey to the end-user.

This module is really the only required module to work with Pyblish, all else is optional. Some studios still prefer to stick with just this core module, as opposed to something like `pyblish-win`, perhaps due to the flexibility of writing their own front-end or because it is lightweight and installable via `pip install pyblish`.

### To the maintainer

This is where GitHub comes in, and this is what the [Developer Guide](https://pyblish.gitbooks.io/developer-guide/content/index.html) is about. To the maintainer, Pyblish is a single module, distributed via `pip install` and comes with (1) integrations, (2) extensions and finally (3) packages. These are meant to be mechanisms that make maintenance and understanding of the overall system easier.

This is where I feel many of your points apply, but please correct me if I'm wrong.

> Instead of a Python package, pyblish-win looks like a distribution for a specific platform to me. 

That's a good point; I think the main reason it includes a Python package is for consistency; every other Pyblish module does this.

> Rather than pyblish-linux, I guess I would expect to see other distributions targeting specific platforms, e.g. pyblish-deb, pyblish-rpm, or pyblish-conda.

I'm all for this. I simply haven't had enough exposure to package creation on Unix platforms to properly distinguish or design for them. At the moment, the Linux distribution merely contains installation instructions, but should host the binary package in which case it makes sense to separate by `deb` and `rpm` etc.

Let me know if you would like to have a go with this, I've been working on [build instructions](https://github.com/pyblish/pyblish-linux/wiki/Installation) for Linux last week, but wouldn't have enough know-how to properly bundle things into a `deb` package.

> For versioning these distributions you could do something similar to what those package management systems do..

That is a good idea!

> Looking at the list of Pyblish repos I might guess that, instead of a distribution targeting the Windows platform, pyblish-win is actually an Integration of Pyblish in to Windows Explorer.

Fair point. How would we better reflect the true nature of this module?

> With the current setup would a user ever write import pyblish_x or import pyblish_win? How about pip install pyblish-win?

Probably not a user. As I mentioned earlier, the Pyblish API is really just `pyblish.api` and is the only publicly facing interface towards the end-user.

The the maintainer, they both expose versions this way, e.g. `pyblish_x.version` and other tools related to their role. In the case of `pyblish_win`, it exposes various tools to bundle the binary files into the resulting `setup.exe`, and I would expect a Debian module to provide the same sort of functionality (if not even an identical interface!).

So, in summary:
- `pyblish` is where everything starts, and contains the full end-user API.
  - `pyblish-qml` and c/o **extend** the core module with e.g. a user-interface
    - `pyblish-win` and c/o **bundle** related modules together into a "distribution"

Let me know if this in any way changes your thoughts about refactoring some of the modules here on GitHub!

**Comment by jedfrechette** [comment]
Nothing wrong with a system that has grown and evolved. That just means it's being used. ;-)

> Technically, the pyblish git repository is a small module upon which all other modules are built. There is only one tool, which is Pyblish. At least this is the impression I would like to convey to the end-user.

I think this is where a little refactoring might help. There seems to be a disconnect between `pyblish` the concrete implementation of a small module and Pyblish the conceptual all in one toolkit; as demoed, for example, in the intro video.

I think Qt is a great example of what I was calling a 'suite'. As you've noted, if you look at the `qt5` super module it doesn't actually implement anything. It delegates all functionality to various submodules. In that sense it is most similar to `pyblish-x` in your current structure, while `pyblish` would be equivalent to something like `qtbase`. I think the benefit to how Qt and Enthought have structured their projects is that it doesn't matter if you're approaching them as a user or a maintainer. Regardless of which hat I have on, when I say Qt it's clear that I'm referring to a specific collection of components, whether or not I know or care how that collection was produced. There doesn't need to be any explicit documentation of the intended design because it is evident in the repository structure, the documentation structure, and even in the installer when it prompts to install the various required and optional components that make up the library.

When I say Pyblish I need to clarify whether I'm talking about little 'p' `pyblish`, the module, or big 'P' Pyblish, the toolkit. This is needed because Pyblish describes a collection, i.e. the top level of a dependency tree, while `pyblish` is at the bottom level of that tree. Qt and `qt5` are both at the top of their dependency tree so no additional clarification is needed in that case.

> This module is really the only required module to work with Pyblish, all else is optional. Some studios still prefer to stick with just this core module, as opposed to something like pyblish-win, perhaps due to the flexibility of writing their own front-end or because it is lightweight and installable via pip install pyblish.

I definitely see the benefits of keeping the core library small and each module independent. Another example would be having an extension like Magenta that might depend on `pyblish-maya` but not `pyblish-nuke`. But that is more or less the same granularity you see when standalone applications build depend on `qtbase` but maybe not `qtsvg`. A developer might install all of `qt5` on their workstation, but it is pretty rare that the application they deliver would specify a dependency against `qt5` with all of its attendant submodules.

> To the maintainer
> 
> This is where GitHub comes in, and this is what the Developer Guide is about. To the maintainer, Pyblish is a single module, distributed via pip install and comes with (1) integrations, (2) extensions and finally (3) packages. These are meant to be mechanisms that make maintenance and understanding of the overall system easier.

Except, of course, right now that's not what you get via pip. You just get the core module.

> > Looking at the list of Pyblish repos I might guess that, instead of a distribution targeting the Windows platform, pyblish-win is actually an Integration of Pyblish in to Windows Explorer.
> 
> Fair point. How would we better reflect the true nature of this module?

I suppose you could use a different naming convention. Maybe something like `distrib-win-pyblish`.

Regarding linux packaging. Although it would be great to have packages for various distros, manual installation is pretty straight-forward on platforms were you can rely on Python 2.7 and PyQT5 already being available, i.e. most current linuxes. Given that `pyblish-linux` pretty much just mirrors the manual installation instructions for `pyblish-x` I wonder if it even needs to exist.

Unfortunately, I'm pretty much stuck working on Windows these days so I'm afraid I wouldn't be a very good person to help with proper packages right now.

**Comment by mottosso** [comment]
@jedfrechette You've got yourself a believer.

Well played! I like it, it makes a lot more sense as `pyblishbase` or `pyblishcore` and to have `pyblish` be the overall package or suite, like Qt. I simply did not see it this way until now.

> When I say Pyblish I need to clarify whether I'm talking about little 'p' pyblish, the module, or big 'P' Pyblish, the toolkit.

This is a great distinction.

> I suppose you could use a different naming convention. Maybe something like distrib-win-pyblish.

Though I'm not a big fan of that particular name, I see what you mean and think you are right.

I'll have a think about how this could remain backwards compatible; seeing as it wouldn't have to change how Pyblish is imported - i.e. `import pyblish` - and the fact that most are using the installer or some derivation of it (which is more or less pyblish-x) I think we should be fine.

Thanks a million for taking the time to think about this, this is big!

**Comment by jedfrechette** [comment]
Glad I could help out a little.

Re: pyblish-win, I'm not a huge fan of the distrib name either but it's the first thing that came to mind. Honestly, I'm not even entirely convinced that it needs changing as it is very much a maintainer only package and it is pretty obvious what it does as soon as you look past the name.

**Comment by mottosso** [comment]
This has been implemented, thanks for the feedback!
- http://forums.pyblish.com/t/renovation-taking-place-today/191/2

Hopefully it makes more sense now, we'll start seeing releases with this new scheme next.
